NFT/tests/harness.rs

```diff
@@ -0,0 +1,633 @@
+
+            let minter = Identity::Address(owner1.wallet.address());
+            let admin = Option::Some(minter.clone());
+            constructor(&deploy_wallet.nft, &admin, true, 1).await;
+
+            mint(&owner1.nft, &minter, 1).await;
+
+            assert_eq!(balance_of(&owner1.nft, &minter).await, 1);
+            assert_eq!(owner_of(&owner1.nft, 1).await, Option::Some(minter.clone()));
+            assert_eq!(approved(&owner1.nft, 1).await, Option::None());
+        }
+
+        #[tokio::test]
+        async fn mints_multiple() {
+            let (deploy_wallet, owner1, _owner2) = setup().await;
+
+            constructor(&deploy_wallet.nft, &Option::None(), false, 10).await;
+
+            let minter = Identity::Address(owner1.wallet.address());
+            mint(&owner1.nft, &minter, 10).await;
+
+            assert_eq!(balance_of(&owner1.nft, &minter).await, 10);
+            for itterator in 1..11 {
+                assert_eq!(
+                    owner_of(&owner1.nft, itterator).await,
+                    Option::Some(minter.clone())
+                );
+                assert_eq!(approved(&owner1.nft, itterator).await, Option::None());
+            }
+        }
+
+        #[tokio::test]
+        async fn does_not_mint_when_not_initalized() {
+            let (deploy_wallet, owner1, _owner2) = setup().await;
+
+            constructor(&deploy_wallet.nft, &Option::None(), false, 1).await;
+
+            let minter = Identity::Address(owner1.wallet.address());
+            mint(&owner1.nft, &minter, 0).await;
+
+            assert_eq!(balance_of(&owner1.nft, &minter).await, 0);
+        }
+    }
+
+    mod reverts {
+
+        use super::*;
+
+        #[tokio::test]
+        #[should_panic(expected = "Revert(42)")]
+        async fn panics_when_not_initalized() {
+            let (_deploy_wallet, owner1, _owner2) = setup().await;
+
+            let minter = Identity::Address(owner1.wallet.address());
+            mint(&owner1.nft, &minter, 1).await;
+        }
+
+        #[tokio::test]
+        #[should_panic(expected = "Revert(42)")]
+        async fn panics_when_minting_more_tokens_than_supply() {
+            let (deploy_wallet, owner1, _owner2) = setup().await;
+
+            constructor(&deploy_wallet.nft, &Option::None(), false, 1).await;
+
+            let minter = Identity::Address(owner1.wallet.address());
+            mint(&owner1.nft, &minter, 2).await;
+        }
+
+        #[tokio::test]
+        #[should_panic(expected = "Revert(42)")]
+        async fn panics_when_minter_does_not_have_access() {
+            let (deploy_wallet, owner1, owner2) = setup().await;
+
+            let minter = Identity::Address(owner2.wallet.address());
+            let admin = Option::Some(Identity::Address(owner1.wallet.address()));
+            constructor(&deploy_wallet.nft, &admin, true, 1).await;
+
+            mint(&owner2.nft, &minter, 1).await;
+        }
+    }
+}
+
+mod set_admin {
```

Comment:
In general, I would match the order of your test cases to the order they are defined in your abi and contract.  I would also order your revert test cases the order in which they appear in your contract (not saying these are out of order just mentioning it)